src/hamiltonian.jl

```diff
@@ -45,9 +45,14 @@ struct PhasePoint{T<:AbstractVecOrMat{<:AbstractFloat}, V<:DualValue}
         @argcheck length(θ) == length(r) == length(ℓπ.gradient) == length(ℓπ.gradient)
         if any(isfinite.((θ, r, ℓπ, ℓκ)) .== false)
             @warn "The current proposal will be rejected due to numerical error(s)." isfinite.((θ, r, ℓπ, ℓκ))
-            E = eltype(T)
-            ℓπ = DualValue(map(v -> isfinite(v) ? v : -E(Inf), ℓπ.value), ℓπ.gradient)
-            ℓκ = DualValue(map(v -> isfinite(v) ? v : -E(Inf), ℓκ.value), ℓκ.gradient)
+            ℓπ = DualValue(
+                map(v -> isfinite(v) ? v : -eltype(T)(Inf), ℓπ.value), 
+                ℓπ.gradient
+            )
+            ℓκ = DualValue(
+                map(v -> isfinite(v) ? v : -eltype(T)(Inf), ℓκ.value), 
+                ℓκ.gradient
+            )
```

Comment:
Maybe I'm missing something, but the new version looks equivalent to that original one?

Reply:
It is exactly the same but it seems that the GPU compiler has some problems inferring that the anonymous function is type-stable if `eltype(T)` is stored in a separate variable. I don't know enough about the GPUCompiler and Julia's type inference whether this is expected behaviour or not.

Reply:
This is interesting to know. 
Can you leave a short note on this so that no one is gonna "optimize" the code to revert this change?

Reply:
So I tried the following
```julia
function my_fun(arr::T) where {T<:Matrix{<:AbstractFloat}}
     E = eltype(T)
     @show Base.return_types(v -> isfinite(v) ? v : eltype(T)(-Inf), (eltype(arr),))
     @show Base.return_types(v -> isfinite(v) ? v : E(-Inf), (eltype(arr),))
end

my_fun(randn(4, 4))
```
which gives
```
Base.return_types((v->begin
            #= REPL[22]:3 =#
            if isfinite(v)
                v
            else
                (eltype(T))(-Inf)
            end
        end), (eltype(arr),)) = Any[Float64]
Base.return_types((v->begin
            #= REPL[22]:4 =#
            if isfinite(v)
                v
            else
                E(-Inf)
            end
        end), (eltype(arr),)) = Any[Any]
```
This suggests that the first version is type-stable, while the second version is not, right? I don't know that much about type stability in Julia so I am not that confident in interpreting these results. Also, I am using Julia 1.6.

Reply:
I agree with your interpretation.
But I don't know why this happens either.
Maybe @mohamed82008 knows?

Reply:
Probably the closure bug.